src/store.js

```diff
@@ -1,61 +1,162 @@
 var {throttle} = require('./lib/util');
+var SimpleSet = require('./lib/simple_set');
 var render = require('./render');
 
 var Store = module.exports = function(ctx) {
+  this._features = {};
+  this._featureIds = new SimpleSet();
+  this._selectedFeatureIds = new SimpleSet();
+  this._changedIds = new SimpleSet();
   this.ctx = ctx;
-  this.features = {};
-  this.featureIds = [];
   this.sources = {
     hot: [],
     cold: []
   };
   this.render = throttle(render, 16, this);
-
   this.isDirty = false;
-  this.changedIds = [];
 };
 
+/**
+ * Sets the store's state to dirty.
+ * @return {Store} this
+ */
 Store.prototype.setDirty = function() {
   this.isDirty = true;
+  return this;
 };
 
-Store.prototype.featureChanged = function(id) {
-  if (this.changedIds.indexOf(id) === -1) {
-    this.changedIds.push(id);
-  }
+/**
+ * Sets a feature's state to changed.
+ * @param {string} featureId
+ * @return {Store} this
+ */
+Store.prototype.featureChanged = function(featureId) {
+  this._changedIds.add(featureId);
+  return this;
 };
 
-Store.prototype.add = function(feature) {
-  this.featureChanged(feature.id);
-  this.features[feature.id] = feature;
-  if (this.featureIds.indexOf(feature.id) === -1) {
-    this.featureIds.push(feature.id);
-  }
-  return feature.id;
+/**
+ * Gets the ids of all features currently in changed state.
+ * @return {Store} this
+ */
+Store.prototype.getChangedIds = function() {
+  return this._changedIds.values();
 };
 
-Store.prototype.get = function(id) {
-  return this.features[id];
+/**
+ * Sets all features to unchanged state.
+ * @return {Store} this
+ */
+Store.prototype.clearChangedIds = function() {
+  this._changedIds.clear();
+  return this;
 };
 
-Store.prototype.getAll = function() {
-  return Object.keys(this.features).map(id => this.features[id]);
+/**
+ * Gets the ids of all features in the store.
+ * @return {Store} this
+ */
+Store.prototype.getAllIds = function() {
+  return this._featureIds.values();
 };
 
-Store.prototype.delete = function (ids) {
+/**
+ * Adds a feature to the store.
+ * @param {Object} feature
+ * @return {Store} this
+ */
+Store.prototype.add = function(feature) {
+  this.featureChanged(feature.id);
+  this._features[feature.id] = feature;
+  this._featureIds.add(feature.id);
+  return this;
+};
+
+/**
+ * Deletes a feature or array of features from the store.
+ * Cleans up after the deletion by deselecting the features.
+ * If changes were made, sets the state to the dirty
+ * and fires an event.
+ * @param {string | Array<string>} featureIds
+ * @return {Store} this
+ */
+Store.prototype.delete = function(featureIds) {
   var deleted = [];
-  ids.forEach((id) => {
-    var idx = this.featureIds.indexOf(id);
-    if (idx !== -1) {
-      var feature = this.get(id);
-      deleted.push(feature.toGeoJSON());
-      delete this.features[id];
-      this.featureIds.splice(idx, 1);
-    }
+  [].concat(featureIds).forEach((id) => {
```

Comment:
Two reasons: So featureIds doesn't have to be an array, can just be a single value (which was preexisting API), and so the array is copied — because if i kept the original sometimes a screwy thing happened that I assume had to do with the array being mutated elsewhere.